README.md
# Avatar Rings

[![style: very good analysis][very_good_analysis_badge]][very_good_analysis_link]
[![Powered by Mason](https://img.shields.io/endpoint?url=https%3A%2F%2Ftinyurl.com%2Fmason-badge)](https://github.com/felangel/mason)
[![License: MIT][license_badge]][license_link]

Custom user avatar widget, waves-like rings around, badges, status points, and many more customizations.

## Preview

![RPReplay_Final1698906431](https://github.com/deam91/avatar_rings/assets/11580629/0b3c0ffc-01ff-4d7f-a6e5-badba8745d6e)
![RPReplay_Final1698906479](https://github.com/deam91/avatar_rings/assets/11580629/588fa0fa-3970-404e-9c43-1b9fa8dbb07b)
![RPReplay_Final1698906588](https://github.com/deam91/avatar_rings/assets/11580629/0654696e-e1a4-4b0e-882a-c7f1b158252b)
![RPReplay_Final1698906656](https://github.com/deam91/avatar_rings/assets/11580629/d8bb7e61-6649-444b-9b34-cf45ea1b5528)
![RPReplay_Final1698906710](https://github.com/deam91/avatar_rings/assets/11580629/b0b8f9bc-e716-4e6f-b8b1-f8676d4300f5)


## Installation 💻

**❗ In order to start using Avatar Rings you must have the [Flutter SDK][flutter_install_link] installed on your machine.**

Install via `flutter pub add`:

```sh
flutter pub add avatar_rings
```

## Usage

```dart
AvatarRings(
  rings: 2, // The number of rings. Defaults to 1.
  radius: 70, // The radius of the entire rings painter. Defaults to 80.
  spaceBetweenRings: 1, // The space between each ring. Defaults to 1.
  fromPeriod: 250, // The start period value. Used to generate a random value starting from this value. Defaults to 150.
  toPeriod: 350, // The end period value. Used to generate a random value ending in this value. Defaults to 200.
  ringsColor: Colors.red, // The color of the rings. Defaults to [Colors.lightBlue].
  ringsColorOpacity: .3, // The opacity of the rings color. Defaults to 0.4.
  shadowColor: Colors.red, // The color of the shadow. Defaults to [Colors.lightBlue].
  shadowColorOpacity: .45, // The opacity of the shadow color. Defaults to 0.4.
  child: Image.asset(
    'assets/person.jpeg',
  ),
),
```

---

## Continuous Integration 🤖

Avatar Rings comes with a built-in [GitHub Actions workflow][github_actions_link] powered by [Very Good Workflows][very_good_workflows_link] but you can also add your preferred CI/CD solution.

Out of the box, on each pull request and push, the CI `formats`, `lints`, and `tests` the code. This ensures the code remains consistent and behaves correctly as you add functionality or make changes. The project uses [Very Good Analysis][very_good_analysis_link] for a strict set of analysis options used by our team. Code coverage is enforced using the [Very Good Workflows][very_good_coverage_link].

---

[flutter_install_link]: https://docs.flutter.dev/get-started/install
[github_actions_link]: https://docs.github.com/en/actions/learn-github-actions
[license_badge]: https://img.shields.io/badge/license-MIT-blue.svg
[license_link]: https://opensource.org/licenses/MIT
[logo_black]: https://raw.githubusercontent.com/VGVentures/very_good_brand/main/styles/README/vgv_logo_black.png#gh-light-mode-only
[logo_white]: https://raw.githubusercontent.com/VGVentures/very_good_brand/main/styles/README/vgv_logo_white.png#gh-dark-mode-only
[mason_link]: https://github.com/felangel/mason
[very_good_analysis_badge]: https://img.shields.io/badge/style-very_good_analysis-B22C89.svg
[very_good_analysis_link]: https://pub.dev/packages/very_good_analysis
[very_good_cli_link]: https://pub.dev/packages/very_good_cli
[very_good_coverage_link]: https://github.com/marketplace/actions/very-good-coverage
[very_good_ventures_link]: https://verygood.ventures
[very_good_ventures_link_light]: https://verygood.ventures#gh-light-mode-only
[very_good_ventures_link_dark]: https://verygood.ventures#gh-dark-mode-only
[very_good_workflows_link]: https://github.com/VeryGoodOpenSource/very_good_workflows


CHANGELOG.md
# 0.1.3

### Features

- add options to set badge text and color
- hide or show the rings option
- option to show the status point and set his color
- option to set a border for the child

# 0.1.2

- add child padding parameter

# 0.1.1

### Features

- add supported/tested platforms
- allow publishing to pub.dev
- add avatar ring previews to the README file

# 0.1.0+12

### Features

- add an optional amount of rings
- add a default radius, rings color and opacity, shadow color and opacity
- add space between rings and start/end period of the waves


test/src/avatar_rings_test.dart
// ignore_for_file: prefer_const_constructors

import 'package:avatar_rings/avatar_rings.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  setUp(() {
    WidgetsFlutterBinding.ensureInitialized();
    WidgetsBinding.instance
      ..resetEpoch()
      ..platformDispatcher.onBeginFrame = null
      ..platformDispatcher.onDrawFrame = null;
  });

  group('AvatarRings', () {
    test('can be instantiated', () {
      expect(
        AvatarRings(
          child: SizedBox(),
        ),
        isNotNull,
      );
    });

    testWidgets('widgets are there', (tester) async {
      const testKey = Key('avatar_rings');
      await tester.pumpWidget(
        MaterialApp(home: const AvatarRings(key: testKey, child: SizedBox())),
      );
      await tester.pump(Duration(milliseconds: 900));

      final ar = find.byKey(Key('avatar_rings'));

      expect(ar, findsOneWidget);
    });
  });
}


lib/src/ring_painter.dart
import 'dart:math';

import 'package:avatar_rings/src/utils.dart';
import 'package:avatar_rings/src/wave.dart';
import 'package:flutter/material.dart';

/// Class used to paint the rings by joining a series of waves for creating a
/// single custom wave, following a circle shape.
class RingPainter extends CustomPainter {
  /// The constructor for the [RingPainter] class
  ///
  /// Accepts an animation [controller], the amount of [rings], the [radius] of
  /// the entire rings painter, a [ringSpace] space between each ring,
  /// a [ringColor] color of the rings, the [ringsColorOpacity] opacity for the
  /// rings, and the [rotation] for each ring.
  ///
  RingPainter({
    required this.controller,
    required this.rings,
    this.radius = 80,
    this.ringSpace = 1.0,
    this.ringColor = Colors.lightBlue,
    this.ringsColorOpacity = .3,
    this.rotation,
  }) : super(repaint: controller);

  /// The animation controller
  final Animation<double> controller;

  /// The radius of the entire rings painter
  final double radius;

  /// The space between each ring
  final double ringSpace;

  /// The color of the rings
  final Color ringColor;

  /// The opacity for the rings color
  final double ringsColorOpacity;

  /// The number of ring waves to paint
  List<List<Wave>> rings = [];

  /// The rotation for each ring
  final double? rotation;

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..color = ringColor.withOpacity(ringsColorOpacity)
      ..style = PaintingStyle.fill
      ..strokeWidth = 2;
    var previousOffset = Offset.zero;
    var savedRadius = radius;
    canvas.save();
    for (var l = 0; l < rings.length; l++) {
      for (var i = -smallDegree; i < pi * 2; i += smallDegree) {
        final x = savedRadius * cos(i + degreeToRadian(rotation!) * l);
        var y = savedRadius * sin(i + degreeToRadian(rotation!) * l);

        for (final wave in rings[l]) {
          y += wave.evaluateSin(x, controller.value * savedRadius);
        }
        final offset = Offset(
          x + size.width / 2,
          y + size.height / 2,
        );
        if (previousOffset == Offset.zero) {
          previousOffset = offset;
        }

        canvas.drawLine(
          previousOffset,
          offset,
          paint,
        );
        previousOffset = offset;
      }
      savedRadius = savedRadius - ringSpace;
      previousOffset = Offset.zero;
    }
    canvas.restore();
  }

  @override
  bool shouldRepaint(RingPainter oldDelegate) => true;
}


lib/src/animated_rings.dart
import 'dart:math';

import 'package:avatar_rings/src/ring_painter.dart';
import 'package:avatar_rings/src/wave.dart';
import 'package:flutter/material.dart';

/// The widget used to animate and style the rings.
class AnimatedRings extends StatefulWidget {
  /// The constructor for the [AnimatedRings] class
  ///
  /// Accepts the amount of [rings], the [radius] of
  /// the entire rings painter, a [spaceBetweenRings] space between each ring,
  /// a [fromPeriod] for the start period value, a [toPeriod] for the end
  /// period value, a [ringsColor] for the color of the rings, and a
  /// [ringsColorOpacity] for the opacity for the rings.
  ///
  /// - The [radius] must be bigger than the [rings] multiplied with the
  /// [spaceBetweenRings].
  ///
  /// - The [rings] must less/equal to 16 to keep performance
  ///
  /// - The [fromPeriod] and [toPeriod] values must be between 10 and 400
  ///
  /// - The [toPeriod] must be bigger or equal to [fromPeriod]
  ///
  const AnimatedRings({
    required this.rings,
    required this.radius,
    required this.spaceBetweenRings,
    required this.fromPeriod,
    required this.toPeriod,
    required this.ringsColor,
    required this.ringsColorOpacity,
    super.key,
  });

  /// The number of rings
  final int rings;

  /// The radius of the entire rings painter
  final double radius;

  /// The start period value. Used to generate a random value starting from [fromPeriod]
  final double fromPeriod;

  /// The start period value. Used to generate a random value ending in [toPeriod]
  final double toPeriod;

  /// The color of the rings
  final Color ringsColor;

  /// The opacity for the rings color
  final double ringsColorOpacity;

  /// The space between each ring
  final double spaceBetweenRings;

  @override
  State<AnimatedRings> createState() => _AnimatedRingsState();
}

class _AnimatedRingsState extends State<AnimatedRings>
    with TickerProviderStateMixin {
  late AnimationController _controller;
  List<List<Wave>> waves = [];
  double? rotation;

  /// Create the wave objects used to build a ring.
  /// The `amplitude`, `period`, and `phase` are random values following the rules:
  ///  - `amplitude` is a random value between 0 and 4
  ///  - `period` is a random value between 100 and 200 by default. Values of `fromPeriod` and `toPeriod`
  ///  - `phase` is by default 2*pi, the phase of a wave
  List<Wave> _buildWaves() {
    final waves = <Wave>[];
    for (var i = 0; i < 5; i++) {
      final amplitude = Random().nextDouble() * 1 + (0.05 * i);
      final period =
          Random().nextInt((widget.toPeriod - widget.fromPeriod).toInt()) +
              widget.fromPeriod;
      final phase = Random().nextDouble() * 2 * pi;
      waves.add(
        Wave(
          amplitude: amplitude,
          period: period,
          phase: phase,
        ),
      );
    }
    return waves;
  }

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      vsync: this,
      duration: const Duration(minutes: 1),
    );
    // Generate waves for each ring
    for (var i = 0; i < widget.rings; i++) {
      waves.add(_buildWaves());
    }
    // If there is more than 1 ring, calculate the rotation for each one
    if (waves.length > 1) {
      rotation = 360 / waves.length;
    }
    _controller.repeat(reverse: true);
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _controller,
      builder: (BuildContext context, Widget? child) => CustomPaint(
        painter: RingPainter(
          controller: _controller,
          rings: waves,
          radius: widget.radius,
          ringSpace: widget.spaceBetweenRings,
          ringColor: widget.ringsColor,
          ringsColorOpacity: widget.ringsColorOpacity,
          rotation: rotation,
        ),
      ),
    );
  }
}


lib/src/avatar_rings.dart
import 'package:avatar_rings/src/animated_rings.dart';
import 'package:flutter/material.dart';

/// Paint the rings by appending a series of different waves into one per ring,
/// and then drawing each ring following a circle shape.
class AvatarRings extends StatelessWidget {
  /// The constructor for the [AvatarRings] class
  ///
  /// Accepts a [child] widget, the amount of [rings], the [radius] of
  /// the entire rings painter, a [spaceBetweenRings] space between each ring,
  /// a [fromPeriod] for the start period value, a [toPeriod] for the end
  /// period value, a [ringsColor] for the color of the rings, a
  /// [ringsColorOpacity] for the opacity for the rings, and the [shadowColor]
  /// and [shadowColorOpacity] for color and opacity of the shadow.
  ///
  /// - The [radius] must be bigger than the [rings] multiplied with the
  /// [spaceBetweenRings].
  ///
  /// - The [rings] must less/equal to 16 to keep performance
  ///
  /// - The [fromPeriod] and [toPeriod] values must be between 10 and 400
  ///
  /// - The [toPeriod] must be bigger or equal to [fromPeriod]
  ///
  const AvatarRings({
    required this.child,
    this.showRings = true,
    this.rings = 4,
    this.radius = 80,
    this.spaceBetweenRings = 1,
    this.fromPeriod = 150,
    this.toPeriod = 200,
    this.ringsColor = Colors.lightBlue,
    this.ringsColorOpacity = .4,
    this.shadowColor = Colors.lightBlue,
    this.shadowColorOpacity = .4,
    this.showStatusPoint = false,
    this.statusPointColor = Colors.green,
    this.padding,
    this.badgeText,
    this.badgeColor,
    this.border,
    super.key,
  })  : assert(
          radius > rings * spaceBetweenRings,
          'The radius must the bigger than the rings multiplied by the space between rings',
        ),
        assert(
          rings <= 10,
          'The rings must less/equal to 16 to keep performance',
        ),
        assert(
          fromPeriod > 10 ||
              toPeriod > 10 ||
              fromPeriod <= 400 ||
              toPeriod <= 400,
          'The fromPeriod and toPeriod values must be between 10 and 400',
        ),
        assert(
          fromPeriod <= toPeriod,
          'The toPeriod must be bigger or equal to fromPeriod',
        );

  /// The child widget for the rings avatar.
  final Widget child;

  /// The badge text widget to show as a notification.
  final Widget? badgeText;

  /// The badge color.
  final Color? badgeColor;

  /// Hide or show the rings. Defaults to true.
  final bool showRings;

  /// The number of rings. Defaults to 1.
  final int rings;

  /// Show the status point at the bottom right corner. Defaults to false.
  final bool showStatusPoint;

  /// The status point color. Defaults to [Colors.green].
  final Color? statusPointColor;

  /// The radius of the entire rings painter. Defaults to 80.
  final double radius;

  /// The border for the child.
  final Border? border;

  /// The padding for the child widget. Defaults to a calculation based on the amount of rings.
  final double? padding;

  /// The start period value. Used to generate a random value starting from [fromPeriod]. Defaults to 150.
  final double fromPeriod;

  /// The end period value. Used to generate a random value ending in [toPeriod]. Defaults to 200.
  final double toPeriod;

  /// The color of the rings. Defaults to [Colors.lightBlue].
  final Color ringsColor;

  /// The opacity of the rings color. Defaults to 0.4.
  final double ringsColorOpacity;

  /// The color of the shadow. Defaults to [Colors.lightBlue].
  final Color shadowColor;

  /// The opacity of the shadow color. Defaults to 0.4.
  final double shadowColorOpacity;

  /// The space between each ring. Defaults to 1.
  final double spaceBetweenRings;

  @override
  Widget build(BuildContext context) {
    final childPadding = padding ?? radius / 10;
    return SizedBox(
      key: const Key('avatar_rings_sizedbox'),
      height: radius * 2,
      width: radius * 2,
      child: Stack(
        key: const Key('avatar_rings_stack'),
        children: [
          Center(
            key: const Key('avatar_rings_shadow'),
            child: Container(
              margin: EdgeInsets.all(rings.toDouble()),
              width: radius * 2,
              height: radius * 2,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                boxShadow: [
                  BoxShadow(
                    color: shadowColor.withOpacity(shadowColorOpacity),
                    spreadRadius: padding != null
                        ? padding!
                        : (showRings ? rings.toDouble() * 2 : 10),
                    blurRadius: padding != null
                        ? padding! * 2
                        : (showRings ? rings.toDouble() * 2 : 10),
                    offset: const Offset(
                      0,
                      1,
                    ), // changes position of shadow
                  ),
                ],
              ),
            ),
          ),
          if (showRings)
            Center(
              key: const Key('avatar_rings_rings'),
              child: AnimatedRings(
                rings: rings,
                radius: radius,
                ringsColor: ringsColor,
                ringsColorOpacity: ringsColorOpacity,
                fromPeriod: fromPeriod,
                toPeriod: toPeriod,
                spaceBetweenRings: spaceBetweenRings,
              ),
            ),
          Center(
            key: const Key('avatar_rings_child'),
            child: Padding(
              padding: EdgeInsets.all(showRings ? childPadding : 0),
              child: DecoratedBox(
                decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  border: border,
                ),
                child: Padding(
                  padding: EdgeInsets.all(showRings ? 5 : 0),
                  child: Stack(
                    children: [
                      ClipOval(
                        child: child,
                      ),
                      if (radius / 2.5 > 10 && showStatusPoint)
                        Positioned(
                          bottom: 0,
                          right: 0,
                          child: DecoratedBox(
                            decoration: ShapeDecoration(
                              color: statusPointColor,
                              shape: const StadiumBorder(),
                            ),
                            child: Padding(
                              padding: EdgeInsets.all(radius / 8),
                            ),
                          ),
                        ),
                      if (badgeText != null)
                        Positioned(
                          top: 0,
                          right: 0,
                          child: DecoratedBox(
                            decoration: ShapeDecoration(
                              color: badgeColor ?? Colors.red,
                              shape: const StadiumBorder(),
                            ),
                            child: Padding(
                              padding: const EdgeInsets.symmetric(
                                horizontal: 8,
                                vertical: 2,
                              ),
                              child: FittedBox(
                                fit: BoxFit.scaleDown,
                                child: radius > 30 ? badgeText : null,
                              ),
                            ),
                          ),
                        ),
                    ],
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
